src/config.rs

```diff
@@ -97,46 +200,65 @@ impl ConfigStore {
     }
 
     /// Returns the setting with the given key. If the setting is not found in the current
-    /// store, it will load the key from the storage. If the key is still not found, it will
+    /// storage, it will load the key from the storage. If the key is still not found, it will
     /// return the default value for the given key. Note that if the key is not found and no
     /// default value is specified, this function will panic.
-    pub fn get(&mut self, key: &str) -> Setting {
-        if !self.has(key) {
-            panic!("Setting {} not found", key);
-        }
-
-        if let Some(setting) = self.settings.get(key) {
-            return setting.clone();
+    pub fn get(&self, key: &str) -> Option<Setting> {
+        if let Some(setting) = self.settings.lock().unwrap().borrow().get(key) {
+            return Some(setting.clone());
         }
 
         // Setting not found, try and load it from the storage adapter
-        if let Some(setting) = self.storage.get_setting(key) {
-            self.settings.insert(key.to_string(), setting.clone());
-            return setting.clone();
+        if let Some(setting) = self.storage.get(key) {
+            self.settings
+                .lock()
+                .unwrap()
+                .borrow_mut()
+                .insert(key.to_string(), setting.clone());
+            return Some(setting.clone());
         }
 
         // Return the default value for the setting when nothing is found
-        let info = self.settings_info.get(key).unwrap();
-        info.default.clone()
+        if let Some(info) = self.settings_info.get(key) {
+            return Some(info.default.clone());
+        }
+
+        // At this point we haven't found the key in the store, we haven't found it in storage, and we
+        // don't have a default value. This is a programming error, so we panic.
+        panic!("config: Setting {} is not known", key);
     }
 
     /// Sets the given setting to the given value. Will persist the setting to the
-    /// storage.
-    pub fn set(&mut self, key: &str, value: Setting) {
-        if !self.has(key) {
-            panic!("key not found");
-        }
-        let info = self.settings_info.get(key).unwrap();
+    /// storage. Note that the setting MUST have a settings-info entry, otherwise
+    /// this function will not store the setting.
+    pub fn set(&self, key: &str, value: Setting) {
+        let info = match self.settings_info.get(key) {
+            Some(info) => info,
+            None => {
+                warn!("config: Setting {} is not known", key);
+                return;
+            }
+        };
+
         if mem::discriminant(&info.default) != mem::discriminant(&value) {
-            panic!("value is of different type than setting expects")
+            warn!(
+                "config: Setting {} is of different type than setting expects",
+                key
+            );
+            return;
```

Comment:
I guess we could try to convert the setting here to the expected type. (e.g int to string)